Depth image

anand2718 opened the pull request:
Added a topic which publishes a depth image of type sensor_msgs/Image along with the disparity.

anandsekar272 committed:
first pass of adding the depth image conversion (untested)

anandsekar272 committed:
Fixed small typo

anandsekar272 committed:
Revert "Fixed small typo"

This reverts commit b26dc825c8bb71ba85e78e92f191398a5c228220.

anandsekar272 committed:
Properly declaring argument

amarburg commented on gpu_stereo_image_proc_common/src/msg_conversions.cpp:
I don't think this will work .. both DisparityImage.data and Image.data are pointers to uint8 (that is, basically pointers to bytes), but the actual data is ... 32 bit floats for both?   So if you're going to step over the array you need to cast it as a float.

Simplest method might be to use cv_bridge and do the work in OpenCV?   A slightly more brittle way might be something like:

```
float *disp_in = disp_msg->data;
float *depth_out = depth_msg->data; 

for( ... ) {
   depth_out[ ... ] = f * T / disp_in[ ... ]
}
```

Should also check that `disp_msg->image.encoding` is `32FC1` otherwise your algorithm will break.   It always will be (we control the incoming format) but it would be easy to change something elsewhere in the code and have this break mysteriously.

amarburg commented on gpu_stereo_image_proc_visionworks/src/nodelet/vx_disparity.cpp:
```diff
@@ -311,15 +317,22 @@ void VXDisparityNodelet::imageCb(const ImageConstPtr &l_image_msg,
           l_image_msg, masked_disparityS16, scaled_model, min_disparity,
           max_disparity, border);
       pub_disparity_.publish(masked_disp_msg);
```
You might as well consolidate all of the `pub_disparity_.publish()` calls down at the end as well.

amarburg commented on gpu_stereo_image_proc_visionworks/src/nodelet/vx_disparity.cpp:
```diff
@@ -311,15 +317,22 @@ void VXDisparityNodelet::imageCb(const ImageConstPtr &l_image_msg,
+    disp_msg_to_convert = disp_msg;
   }
 
+  // Publish a depth image of type sensor_msgs/Image
```
For efficiency, can we add a `numSubscribers()` check and only publish a depth image if needed?

amarburg reviewed (changes requested):
Looks good,  see comments in the DepthImage conversion function....

anandsekar272 committed:
Preliminary commit to see if class builds on Trisect (untested)

anandsekar272 committed:
Fixing typo (untested)

anandsekar272 committed:
Adding return types to getter definitions

amarburg commented:
Cool.  Let me know when this is stable and I'll review it again...

anandsekar272 committed:
Replacing VisionWorks calls to disparity and depth with new objects (untested)

anandsekar272 committed:
Fixing typo

anandsekar272 committed:
Using the correct arguments for each kind of disparity call

amarburg commented on gpu_stereo_image_proc_common/include/gpu_stereo_image_proc/msg_conversions.h:
```diff
@@ -5,9 +5,26 @@
+    sensor_msgs::ImagePtr depth;
+};
+
 stereo_msgs::DisparityImagePtr
```
I **think** you can remove the standalone functions `disparityToDisparityImage` and `disparityImageToDepthImage`?

anand2718 replied:
If we're only using the VisionWorks part of the code (not libsgm or opencv parts), then we should remove it for sure. Otherwise, I'd have to replace calls to these functions in other parts of the code, and I'm not sure I can test if those end up working (I have libsgm skiplisted when building currently, I'm unsure if I can build it for testing if I do mess with the libsgm code).

amarburg replied:
Oh yeah.   Forgot about that.   OK, leave it in place but add a \todo that we can remove them if/when the libsgm and opencv modules have been ported to use the class.    Long term, might as well port them because it's straightforward and then they can publish depth images.

amarburg commented on gpu_stereo_image_proc_common/src/msg_conversions.cpp:
```diff
@@ -2,10 +2,91 @@
+  disp_msg->delta_d = inv_dpp;
+
+  // Create the depth image
+  ImagePtr depth_msg = boost::make_shared<Image>();
```
Since we always use the disparity image, but don't always use the depth image (like for the debug disparity images), the Depth image creation should be just-in-time in the getDepth()  function.   

Save whatever state you need to do that calculation ... I think you have most if not all of it in the disparity message.

FWIW I don't think you need to store the Depth image but I don't think it makes a huge difference.

anand2718 replied:
This is a good optimization - I can easily save the intermediate Mat and move the processing to getDepth(). I'll focus on making sure the depth image is correct before doing this.

amarburg commented on gpu_stereo_image_proc_common/src/msg_conversions.cpp:
```diff
@@ -2,10 +2,91 @@
+  depth_msg->header = disp_msg->image.header;
+  depth_msg->height = dimage.height;
+  depth_msg->width = dimage.width;
+  depth_msg->encoding = dimage.encoding;
```
Just to be pedantic, set `depth_msg->encoding` to `sensor_msgs::image_encodings::TYPE_32FC1`.   Yes, that is `dimage`'s type, but in this case we're making an affirmative decision for the depth to be floats ... we aren't copying `dimage`'s type.

amarburg replied:
Then you don't need the assert that follows...

amarburg commented on gpu_stereo_image_proc_common/src/msg_conversions.cpp:
```diff
@@ -2,10 +2,91 @@
+  depth_msg->data.resize(dimage.step * dimage.height);
+  cv::Mat_<float> depth_mat(dimage.height, dimage.width, (float *)&depth_msg->data[0],
+                       dimage.step);
+  cv::divide((disp_msg->f * disp_msg->T), dmat, depth_mat);
```
Heck yeah.   (oh yeah, if you move this to the `getDepth` function you'll need to re-create `dmat` (copy line 29) but that's no big deal.

amarburg commented on gpu_stereo_image_proc_visionworks/src/nodelet/vx_disparity.cpp:
```diff
@@ -307,19 +313,27 @@ void VXDisparityNodelet::imageCb(const ImageConstPtr &l_image_msg,
 
+  // Publish a depth image of type sensor_msgs/Image
+  ImagePtr depth_msg = disparityImageToDepthImage(disp_msg);
+  pub_depth_.publish(depth_msg);
```
why isn't this `dg_for_depth.getDepth()`?

amarburg reviewed (commented):
Looks good.  See the comments below:

1. Remove the redundant stand-alone `disparityTo...` functions?
2. Move the Depth image creation to be just-in-time in `getDepth`
3. Strangely you aren't actually calling `getDepth` in the image callback

amarburg committed:
Improvements to depth image generation.

- Split depth image generation out from DisparityImageGenerator
constructor, so it only happens if needed.
- Added threshold to discard disparities less than minimum
- Refactored confidence threshold to make code flow cleaner.

amarburg commented:
OK, check my latest changes to the branch.  Mostly just cleanup. 

I think it's working ... but strangely I can't view the depth image (just comes up black).   minMaxLoc() does report a minimum depth value of ~0.13 and max of ~42-ish.     I'll work on visualizing the depth, but can you run the code and see how it looks in your visualization tools.

amarburg committed:
Explicitly set all depths calculated from bad disparities to NaN.

amarburg committed:
Merge remote-tracking branch 'origin/melodic-devel' into depth_image
